examples/tcp-window-clamps.bpf.c

```diff
@@ -0,0 +1,90 @@
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_core_read.h>
+#include "maps.bpf.h"
+
+/* Minimum value for tp->rcv_ssthresh that is not considered a clamp */
+#define MIN_CLAMP 32 * 1024
+
+struct {
+    __uint(type, BPF_MAP_TYPE_ARRAY);
+    __uint(max_entries, 1);
+    __type(key, u32);
+    __type(value, u64);
+} tcp_window_clamps_total SEC(".maps");
+
+static int handle_tcp_sock(struct tcp_sock *tp)
+{
+    u32 zero = 0, rcv_ssthresh;
+
+    if (!tp) {
+        return 0;
+    }
+
+    rcv_ssthresh = BPF_CORE_READ(tp, rcv_ssthresh);
+
+    if (rcv_ssthresh < MIN_CLAMP) {
+        increment_map(&tcp_window_clamps_total, &zero, 1);
+    }
+
+    return 0;
+}
+
+#ifdef FENTRY_SUPPORT
+// If fentry/fexit is supported, use it for simpler and faster probe.
+// You need to pass -DFENTRY_SUPPORT in compiler flags to enable this.
+
+SEC("fexit/tcp_try_rmem_schedule")
+int BPF_PROG(tcp_try_rmem_schedule_exit, struct sock *sk)
+{
+    return handle_tcp_sock((struct tcp_sock *) sk);
+}
+
+#else
+// Otherwise, fall back to good old kprobe.
+
+struct {
+    __uint(type, BPF_MAP_TYPE_LRU_HASH);
+    __uint(max_entries, 1024);
+    __type(key, u64);
+    __type(value, struct sock *);
+} tcp_rmem_schedule_enters SEC(".maps");
+
+static int enter_key()
+{
+    u64 pid = bpf_get_current_pid_tgid();
+    if (pid) {
+        return pid;
+    }
+
+    return bpf_get_smp_processor_id();
```

Comment:
do we care about `proc_id()` and `pid` collision in the key?

Reply:
It's unlikely to happen in practice, but probably best to guard against it just in case.